src/persist-client/src/read.rs

```diff
@@ -98,18 +140,57 @@ where
     pub async fn poll_next(
         &mut self,
         timeout: Duration,
-    ) -> Result<ListenEvent<K, V, T, D>, LocationError> {
-        todo!("{:?}", timeout)
+    ) -> Result<Vec<ListenEvent<K, V, T, D>>, LocationError> {
+        trace!("Listen::poll_next timeout={:?}", timeout);
+        let mut ret = Vec::new();
+        loop {
+            let state = self.shard.clone().into_inner();
+            let state = state.lock().await;
+            let new_upper = Antichain::from(
+                state
+                    .upper
+                    .iter()
+                    .map(|x| T::decode(*x))
+                    .collect::<Vec<_>>(),
+            );
+            if PartialOrder::less_than(&self.upper, &new_upper) {
+                // TODO: We could order contents to avoid scanning the entire
+                // thing every time, but this impl is meant as a placeholder, so
+                // let's see how far we get without that.
+                let mut updates = Vec::new();
+                for (k, v, t, d) in state.contents.iter() {
+                    let t = T::decode(*t);
+                    if self.as_of.less_than(&t) && self.upper.less_equal(&t) {
+                        debug_assert_eq!(new_upper.less_equal(&t), false);
+                        let k = K::decode(&k);
+                        let v = V::decode(&v);
+                        let d = D::decode(*d);
+                        updates.push(((k, v), t, d));
+                    }
+                }
+                if !updates.is_empty() {
+                    ret.push(ListenEvent::Updates(updates));
+                }
+                self.upper = new_upper;
+                ret.push(ListenEvent::Progress(self.upper.clone()));
+                return Ok(ret);
+            }
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        }
     }
 }
 
 /// A "capability" granting the ability to read the state of some shard at times
 /// greater or equal to `self.since()`.
+#[derive(Debug)]
 pub struct ReadHandle<K, V, T, D>
 where
     T: Timestamp + Lattice + Codec64,
 {
-    _phantom: PhantomData<(K, V, T, D)>,
+    pub(crate) reader_id: ReaderId,
+    pub(crate) state: Shard<K, V, T, D>,
+
+    pub(crate) since: Antichain<T>,
 }
 
 impl<K, V, T, D> ReadHandle<K, V, T, D>
```

Comment:
does it make sense to make this a trait? then we can keep the in-mem impl when we switch to the persisted one?

Reply:
we'll still have an in-mem impl behind a trait abstraction, it'll just be at a different level than this! very similar to the current `mz_persist::storage::Blob` if you're curious

Reply:
Ill need to read into `Blob`!